File: stack/106. 从中序与后序遍历序列构造二叉树.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
 var buildTree = function(inorder, postorder) {
    // 解题思路：通过后序遍历找到根节点值创建根节点
    // 
    // 找到中序遍历的左子树，中序遍历的右子树， 后序遍历的左子树，后序遍历的右子树
    // 递归调用--通过左子树中序遍历,后序遍历子树,找到左子树根节点递归调用

    if(!inorder.length) return null;
    let val = postorder[postorder.length - 1];
    let root = new TreeNode(val);
    let per = 0;
    while(per < inorder.length){
        if(val == inorder[per]){
            break;
        }
        per++;
    }
    let l_in = [];
    let l_post = [];
    let r_in = []; 
    let r_post = [];
    for(let i = 0 ; i < per; i++){
        l_in.push(inorder[i]);
        l_post.push(postorder[i])
    }
    for(let i = per; i < inorder.length - 1;i++){
        r_in.push(inorder[i+1]);
        r_post.push(postorder[i])
    }
    root.left = buildTree(l_in,l_post);
    root.right = buildTree(r_in,r_post);
    
    return root

};

File: Python/树/106. 从中序与后序遍历序列构造二叉树.py
# -*- codeing = utf-8 -*-
# @Time : 2023-9-8 21:48
# @Author : 刘永奇
# @File : 106. 从中序与后序遍历序列构造二叉树.py
# @Software : PyCharm
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if len(postorder) == 0: return None
        l = len(postorder)
        pos = 0
        while inorder[pos] != postorder[-1]:pos+=1
        l_in = inorder[:pos]
        l_post = postorder[:pos]
        r_in = inorder[pos+1:]
        r_post = postorder[pos:l-1]
        node = TreeNode(postorder[-1])
        node.left = self.buildTree(l_in,l_post)
        node.right = self.buildTree(r_in,r_post)
        return node